core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,566 @@
+     * If true is returned, ctx.entryIndex keeps the index of the found entry
+     * and ctx.entryState keeps the state.
+     *
+     * IMPORTANT:
+     *  1. Throws IllegalStateException if too much collisions are found
+     *
+     * @param ctx the context that will follow the operation following this key allocation
+     * @param key the key to write
+     * @param hashIdx hashIdx=fullHash||bit_size_mask
+     * @param fullHashIdx fullHashIdx=hashFunction(key)
+     * @return true only if the key is found (ctx.entryState keeps more details).
+     *         Otherwise (false), key wasn't found, ctx with it's buffers is invalidated
+     */
+    boolean lookUp(ThreadContext ctx, K key, int hashIdx, long fullHashIdx) {
+        // start from given hash index
+        // and check the next `collisionEscapes` indexes if previous index is occupied
+        int currentLocation = hashIdx;
+        int collisionEscapesLocal = collisionEscapes.get();
+
+        // as far as we didn't check more than `collisionEscapes` indexes
+        while (Math.abs(currentLocation - hashIdx) < collisionEscapesLocal) {
+
+            ctx.entryIndex = currentLocation; // check the entry candidate
+            // entry's key is read into ctx.tempKey as a side effect
+            ctx.entryState = getEntryState(ctx, currentLocation, key, fullHashIdx);
+
+            // value and key slices are read during getEntryState() unless the entry is
+            // fully deleted (EntryState.DELETED) in this case we cannot compare the key (!)
+            // also deletion linearization point is checked during getEntryState()
+            if (ctx.entryState != EntryState.DELETED &&
+                equalKeyAndEntryKey(ctx.key, key, ctx.entryIndex, fullHashIdx)) {
+                // EntryState.VALID --> the key is found
+                // DELETED_NOT_FINALIZED/UNKNOWN --> key doesn't exists
+                //                      and there is no need to continue to check next entries
+                // INSERT_NOT_FINALIZED --> before linearization point, key doesn't exist
+                // when more than unique keys can be concurrently inserted, need to check further!
+                return (ctx.entryState == EntryState.VALID) ? true : false;
+            }
+            // not in this entry, move to next
+            currentLocation = (currentLocation + 1) % entriesCapacity; // cyclic increase
+        }
+        ctx.invalidate();
+        return false;
+    }
+
+
+    /**
+     * findSuitableEntryForInsert finds the entry where given hey is going to be inserted.
+     * Given initial hash index for the key, it checks entries[hashIdx] first and continues
+     * to the next entries up to 'collisionEscapes', if the previous entries are all occupied.
+     * If true is returned, ctx.entryIndex keeps the index of the chosen entry
+     * and ctx.entryState keeps the state.
+     *
+     * IMPORTANT:
+     *  1. Throws IllegalStateException if too much collisions are found
```

Comment:
`IllegalStateException` is thrown:
1. ctx.entryState has unexpected value. This would be a bug. For example if new state is added and this coded is not fixed appropriately.
2. Collision chain length is found be increased too much. It is better to get noticed about this situation, as this can make the user to consider another hash function. If assert would be used it would be compiled out soon.

Reply:
If it is for debugging, use `assert`.

Reply:
This is a point I would discuss, once we collect all the review comments.

Reply:
Change to print

Reply:
done